### PR TITLE
longhorn: implement the dedicated recovery backend

### PR DESCRIPTION
**Pull request by derekbit**
https://github.com/longhorn/longhorn/issues/2293

Signed-off-by: Derek Su <derek.su@suse.com>

### CODE REVIEW

File: src/SAL/recovery/recovery_longhorn.c [hunk 1]
```diff
@@ -0,0 +1,766 @@
...
+	err = gethostname(host, sizeof(host));
+	if (err) {
+		LogEvent(COMPONENT_CLIENTID,
+				 "Failed to gethostname: %s (%d)",
+				 strerror(errno), errno);
+		return;
+	}
+
+	longhorn_create_clid_name(clientid);
+	encoded_cid_recov_tag = url_encode(clientid->cid_recov_tag);
```
Comment (innobead) [inline src/SAL/recovery/recovery_longhorn.c]: I am curious why not just use `curl_easy_escape`? in particular, you already use curl lib here. You can remove some repeated code here for `url_encode`.

Comment (derekbit) [reply src/SAL/recovery/recovery_longhorn.c]: Ah, thanks for the hint. I was not aware of the curl API.

Comment (derekbit) [reply src/SAL/recovery/recovery_longhorn.c]: Replaced with `curl_easy_escape()`. Thank you.

File: src/SAL/recovery/recovery_longhorn.c [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+{
+	char host[NI_MAXHOST];
+	char payload[PAYLOAD_MAX];
+	char *response = NULL;
+	size_t response_size = 0;
+	char *version = NULL;
+	int err = 0;
+	int res = 0;
+
+	err = gethostname(host, sizeof(host));
```
Comment (innobead) [inline src/SAL/recovery/recovery_longhorn.c:402]: Will `host` keep the same after the share manager pod failed over?

**Review by innobead — COMMENTED** [review]
In general, LGTM. Just one comment waiting to resolve.

File: src/SAL/recovery/recovery_longhorn.c (hunk 2 above)
Comment (derekbit) [reply src/SAL/recovery/recovery_longhorn.c:402]: Yes,
The hostname inside the share-manager pod is same as the name of the share-manager pod.
It is not changed after failover.

**Review by innobead — APPROVED** [review]

File: src/SAL/recovery/recovery_longhorn.c (hunk 2 above)
Comment (innobead) [reply src/SAL/recovery/recovery_longhorn.c:402]: ref: https://github.com/longhorn/longhorn-manager/blob/c3e9b1c8c7d1043134bfbf2c8ceb25c789935bc0/controller/share_manager_controller.go#L784-L784

**Pull request merged**